experimental/framework/fluid-llm/src/shared-tree-diff/sharedTreeDiff.ts

```diff
@@ -0,0 +1,424 @@
+import { type TreeArrayNode, NodeKind } from "@fluidframework/tree";
+
+import { isTreeMapNode } from "./utils.js";
+
+/**
+ * Represents a path through a tree of objects.
+ * number values represent array indices whereas string values represent object keys.
+ */
+export type ObjectPath = (string | number)[];
+
+/**
+ * Represents a create operation between two branches of a tree.
+ * Meaning that an attribute (a shared tree node) was identified as being created.
+ */
+export interface DifferenceCreate {
+	type: "CREATE";
+	path: ObjectPath;
+	value: unknown;
+}
+
+/**
+ * Represents a remove operation between two branches of a tree.
+ * Meaning that an attribute (a shared tree node) was identified as being deleted.
+ * When using object ids, removes are idenitified by an object with a given id no longer existing.
+ */
+export interface DifferenceRemove {
+	type: "REMOVE";
+	path: ObjectPath;
+	oldValue: unknown;
+}
+
+/**
+ * Represents a change operation between two branches of a tree.
+ * Meaning that an attribute (a shared tree node) was identified as being changed from one value to another.
+ */
+export interface DifferenceChange {
+	type: "CHANGE";
+	path: ObjectPath;
+	value: unknown;
+	oldValue: unknown;
+}
+
+/**
+ * Represents a move operation between two branches of a tree.
+ * Meaning that an object (shared tree node) was identified as being moved from one index to another based on its unique id.
+ */
+export interface DifferenceMove {
+	type: "MOVE";
+	path: ObjectPath;
+	newIndex: number;
+	value: unknown;
+}
+
+/**
+ * Union for all possible difference types.
+ */
+export type Difference =
+	| DifferenceCreate
+	| DifferenceRemove
+	| DifferenceChange
+	| DifferenceMove;
+
+interface Options {
+	cyclesFix: boolean;
```

Comment:
Presumably this has a 1:1 definition with zod's `cyclesFix` property? Our trees don't support cycles, so is this needed? If so, docs would be good :)